### PR TITLE
lb4: rework Testing your application

### PR DESCRIPTION
**Pull request by bajtos**
"Testing your application" and "Testing-Your-application-advanced" are behind "Thinking in LoopBack", they are also partially duplicating the content in that page. This pull request contains a rework/rewrite of these two "Testing" pages. See https://github.com/strongloop/loopback-next/issues/553 for more details.

Connect to strongloop/loopback-next#553

In contrary to Acceptance Criteria described in the lb-next issue, I decided to keep all code examples in Thinking in LoopBack without changes. IMO, these examples are important to keep "Thinking in LoopBack" as a self-contained guide that can be read on its own, without having to jump to different pages and back.

This is work-in-progress and should not be landed yet.

**Tasks**

- Write individual sections
  - [x] Project setup
  - [x] Data handling
  - Unit testing  
    - [x] Use test doubles
    - [x] Unit-test your Controllers
    - [x] Unit-test your Models and Repositories
    - [x] Unit-test your Sequence
  - Integration testing
    - [x] Test your Repositories against a real database
    - [x] Test your Controllers and Repositories together
  - Acceptance testing
    - [x] Validate your OpenAPI specification
    - [x] Perform an auto-generated smoke test of your REST API
    - [x] Test your individual REST API endpoints
    - [x] Test Sequence customizations
- [x] "Thinking in LoopBack" - add references to "Testing your application"
- [x] Move useful bits from legacy content from "Testing your application" to the new content and/or "Thinking in LoopBack"
- [x] Move useful bits from "Testing your app Advanced" to the new content and/or "Thinking in LoopBack". Delete the page "Testing your app Advanced".

### CODE REVIEW

File: pages/en/lb4/Testing-Your-application.md [hunk 1]
```diff
@@ -8,39 +8,236 @@ permalink: /doc/en/lb4/Testing-your-application.html
...
+## Overview
 
-### With LoopBack CLI
+An automated test suite is important to ensure your application works as expected, prevent regressions as new features are added and give you confidence to refactor your codebase to keep it easy to further modify.
+
+Thorough automated testing:
+- Prevents regressions from new features/bug fixes (code introductions and refactorings) (after deployment)
+- Helps new and existing developers understand different parts of the codebase (knowledge sharing)
+- Speeds up development over the long run (the code writes itself!)
+- And much more. See [benefits of software testing](http://lmgtfy.com/?q=benefits+of+software+testing)
```
Comment (kjdelisle) [inline pages/en/lb4/Testing-Your-application.md:19]: Shouldn't be a bullet point if it's a continuation of the thought and not the list.

Comment (kjdelisle) [reply pages/en/lb4/Testing-Your-application.md:19]: As much as I find it hilarious to use LMGTFY, I don't think we should be passive-aggressive about helping people see the value of test-driven development. :P

File: pages/en/lb4/Testing-Your-application.md [hunk 2]
```diff
@@ -8,39 +8,236 @@ permalink: /doc/en/lb4/Testing-your-application.html
...
+" %}
+
+The rest of this page contains a reference manual explaining how to write different types of tests.
+
+## Project setup
+
+An automated test suite requires a test runner - a tool that can execute all your tests and produce a test report. We use and recommend [Mocha](https://mochajs.org).
+
+In addition to a test runner, tests need various utilities like an assertion library (we recommend [Should.js](https://shouldjs.github.io)), a library for making HTTP calls and verifying their results (we recommend [supertest](https://github.com/visionmedia/supertest)), a library for creating Test doubles (we recommend [Sinon.JS](http://sinonjs.org/)). While it's possible to install all these tools individually, we found that a good setup requires additional glue to integrate all these individual tools into something that's easy to use. Especially when writing the tests in TypeScript, due to relatively lower quality of official type definitions. We have addressed all these needs by creating our own swiss-army-knife module for testing called [@loopback/testlab](https://www.npmjs.com/package/@loopback/testlab).
```
Comment (kjdelisle) [inline pages/en/lb4/Testing-Your-application.md:36]: Break this line into multiple lines.

Comment (kjdelisle) [reply pages/en/lb4/Testing-Your-application.md:36]: `a library for creating Test doubles...` -- lower-case the word "Test".

Comment (kjdelisle) [reply pages/en/lb4/Testing-Your-application.md:36]: `While it's possible to install all these tools individually, we found that a...`
"we've found that a..."

File: pages/en/lb4/Testing-Your-application.md [hunk 3]
```diff
@@ -8,39 +8,236 @@ permalink: /doc/en/lb4/Testing-your-application.html
...
+- [Test spies](http://sinonjs.org/releases/v4.0.1/spies/) are functions that record arguments, return value, the value of `this` and exception thrown (if any) for all its calls. There are two types of spies: Some are anonymous functions, while others wrap methods that already exist in the system under test.
+
+- [Test stubs](http://sinonjs.org/releases/v4.0.1/stubs/) are functions (spies) with pre-programmed behavior. As spies, stubs can be either anonymous, or wrap existing functions. When wrapping an existing function with a stub, the original function is not called.
+
+- [Test mocks](http://sinonjs.org/releases/v4.0.1/mocks/)  (and mock expectations) are fake methods (like spies) with pre-programmed behavior (like stubs) as well as pre-programmed expectations. A mock will fail your test if it is not used as expected.
+
+{% include note.html content="We recommend against using test mocks. With test mocks, the expectations must be defined before the tested scenario is executed, which breaks the recommended test layout 'arrange-act-assert' (or  'given-when-then') and produces code that's difficult to comprehend.
+" %}
+
+#### Create a stub Repository
```
Comment (kjdelisle) [inline pages/en/lb4/Testing-Your-application.md:202]: Stub

File: pages/en/lb4/Testing-Your-application.md [hunk 4]
```diff
@@ -8,39 +8,236 @@ permalink: /doc/en/lb4/Testing-your-application.html
...
+
+When writing an application accessing data in a database, it's a best practice to use [Repositories](./Repositories.html) to encapsulate all data-access/persistence-related code and let other parts of the application (typically [Controllers](./Controllers.html)) to depend on these Repositories for data access. In order to test Repository dependants (e.g. Controllers) in isolation, we need to provide a test double, usually as a test stub.
```
Comment (kjdelisle) [inline pages/en/lb4/Testing-Your-application.md:204]: Break up into multiple lines.

Comment (kjdelisle) [reply pages/en/lb4/Testing-Your-application.md:204]: `let other parts of the application (typically [Controllers](./Controllers.html)) to depend on these Repositories...`
should be
`let other parts of the application (typically [Controllers](./Controllers.html)) depend on these Repositories...`

File: pages/en/lb4/Testing-Your-application.md [hunk 5]
```diff
@@ -8,39 +8,236 @@ permalink: /doc/en/lb4/Testing-your-application.html
...
+
+In traditional object-oriented languages like Java or C#, in order to allow the unit tests to provide a custom implementation of the repository API, the controller needs to depend on an interface describing the API, and the repository implementation needs to implement this interface. The situation is easier in JavaScript and TypeScript. Thanks to the dynamic nature of the language, it’s possible to mock/stub entire classes.
```
Comment (kjdelisle) [inline pages/en/lb4/Testing-Your-application.md:206]: Break this up into multiple lines.

File: pages/en/lb4/Testing-Your-application.md [hunk 6]
```diff
@@ -8,39 +8,236 @@ permalink: /doc/en/lb4/Testing-your-application.html
...
+
+Creating a test double for a repository class is very easy with Sinon.JS' utility function `createStubInstance`. It's important to create a new stub instance for each unit test in order to prevent unintended re-use of pre-programmed behaviour between (unrelated) tests.
```
Comment (kjdelisle) [inline pages/en/lb4/Testing-Your-application.md:208]: Break this up into multiple lines.

File: pages/en/lb4/Testing-Your-application.md [hunk 7]
```diff
@@ -8,39 +8,236 @@ permalink: /doc/en/lb4/Testing-your-application.html
...
+
+  // your unit tests
+
+  function givenStubbedRepository() {
+    repository = sinon.createStubInstance(ProductRepository);
+  }
+});
+```
+
+In your unit-tests, you will usually want to program the behavior of stubbed methods (what should they return) and then verify that the Controller (unit under test) called the right method with the correct arguments.
```
Comment (kjdelisle) [inline pages/en/lb4/Testing-Your-application.md]: unnecessary hyphen in `unit-tests`.

Comment (kjdelisle) [reply pages/en/lb4/Testing-Your-application.md]: `(what should they return)` -> `(what they should return)`

File: pages/en/lb4/Testing-Your-application.md [hunk 8]
```diff
@@ -8,39 +8,236 @@ permalink: /doc/en/lb4/Testing-your-application.html
...
+
+Configure stub's behavior at the beginning of your unit-test (in the "arrange" or "given" section):
```
Comment (kjdelisle) [inline pages/en/lb4/Testing-Your-application.md]: `unit-test` -> `unit test`

File: pages/en/lb4/Testing-Your-application.md [hunk 9]
```diff
@@ -8,39 +8,236 @@ permalink: /doc/en/lb4/Testing-your-application.html
...
+
+```ts
+// repository.find() will return a promise that
+// will be resolved with the provided array
+const findStub = repository.find as sinon.SinonStub;
+findStub.resolves([{id: 1, name: 'Pen'}]);
+```
+
+Verify how was the stubbed method executed at the end of your unit-test (in the "assert" or "then" section):
```
Comment (kjdelisle) [inline pages/en/lb4/Testing-Your-application.md]: `unit-test` -> `unit test`

File: pages/en/lb4/Testing-Your-application.md [hunk 10]
```diff
@@ -8,39 +8,236 @@ permalink: /doc/en/lb4/Testing-your-application.html
...
+
+See [Unit test your controllers](#unit-test-your-controllers) for a full example.
+
+#### Create a stub Service
+
+To be done. The initial Beta release does not include Services as a first-class feature.
+
+### Unit-test your Controllers
+
+A typical unit-test creates a controller instance with dependencies replaced by Test doubles and directly calls the tested method. In the example below, we give the controller a stub implementation of its repository dependency, see [Create a stub repository](#create-a-stub-repository) for a detailed explanation.
```
Comment (kjdelisle) [inline pages/en/lb4/Testing-Your-application.md]: `unit-test` -> `unit test`

Comment (kjdelisle) [reply pages/en/lb4/Testing-Your-application.md]: `Test doubles` -> `test doubles`

File: pages/en/lb4/Testing-Your-application.md [hunk 11]
```diff
@@ -8,39 +8,236 @@ permalink: /doc/en/lb4/Testing-your-application.html
...
+    });
+  });
+
+  function givenStubbedRepository() {
+    repository = sinon.createStubInstance(ProductRepository);
+  }
+});
+```
+
+### Unit-test your Models and Repositories
```
Comment (kjdelisle) [inline pages/en/lb4/Testing-Your-application.md]: `Unit-test your Models and Repositories` -> `Unit Testing Models and Repositories`

File: pages/en/lb4/Testing-Your-application.md [hunk 12]
```diff
@@ -8,39 +8,236 @@ permalink: /doc/en/lb4/Testing-your-application.html
...
+
+In a typical LoopBack appliction, models and repositories are relying mostly on the behavior provided by the framework (`@loopback/repository` package) and there is no need to test the LoopBack's built-in functionality unit-test level (the functionality has been already verified by the tests in the framework). What does need new unit-tests is any additional application-specific API.
```
Comment (kjdelisle) [inline pages/en/lb4/Testing-Your-application.md]: Break up into multiple lines.

Comment (kjdelisle) [reply pages/en/lb4/Testing-Your-application.md]: `appliction` -> `application`

Comment (kjdelisle) [reply pages/en/lb4/Testing-Your-application.md]: `...and there is no need to test the LoopBack's built-in functionality unit-test level...`
Awkward phrasing. I'd probably rephrase it to be direct about not needing to unit test anything we provide:
```
Your unit tests should not attempt to validate classes or functions provided by the
`@loopback/repository` (ex. EntityCrudRepository).
Instead, use test doubles to stub or mock these classes and methods.
```

File: pages/en/lb4/Testing-Your-application.md [hunk 13]
```diff
@@ -8,39 +8,236 @@ permalink: /doc/en/lb4/Testing-your-application.html
...
+
+For example, if our `Person` Model has properties `firstname`, `middlename` and `surname` and provides a function to obtain the full name, then we should write unit tests to verify the implementation of this additional method.
+
+```ts
+// test/unit/models/person.model.unit.ts
+import {Person} from '../../models/person.model.ts'
+import {expect} from '@loopback/testlab';
+
+describe('Person (unit)', () => {
+  // we recommend to group tests by method names
```
Comment (kjdelisle) [inline pages/en/lb4/Testing-Your-application.md]: `we recommend grouping tests by method names`

File: pages/en/lb4/Testing-Your-application.md [hunk 14]
```diff
@@ -8,39 +8,236 @@ permalink: /doc/en/lb4/Testing-your-application.html
...
+    });
+
+    it('omits middlename when not present', () => {
+      // etc.
+    });
+  });
+});
+```
+
+Writing a unit test for a custom Repository methods is not straightforward because `CrudRepositoryImpl` is based on legacy loopback-datasource-juggler that was not designed with Dependency Injection in mind. We are recommending to use integration tests to verify the implementation of custom Repository methods, see [Test your Repositories against a real database](#test-your-repositories-against-a-real-database) in [Integration Testing](#integration-testing).
```
Comment (kjdelisle) [inline pages/en/lb4/Testing-Your-application.md]: Break up into separate lines.

Comment (kjdelisle) [reply pages/en/lb4/Testing-Your-application.md]: `Writing a unit test for a custom Repository methods`
should either lose the plurality on "methods" or remove the leading "a" from "a custom".

File: pages/en/lb4/Testing-Your-application.md [hunk 15]
```diff
@@ -8,39 +8,236 @@ permalink: /doc/en/lb4/Testing-your-application.html
...
+
+### Unit-test your Sequence
```
Comment (kjdelisle) [inline pages/en/lb4/Testing-Your-application.md]: `Unit Testing Sequences`

File: pages/en/lb4/Testing-Your-application.md [hunk 16]
```diff
@@ -8,39 +8,236 @@ permalink: /doc/en/lb4/Testing-your-application.html
...
+
+While it's possible to test a custom Sequence class in isolation, we think it's better to rely on acceptance-level tests in this exceptional case. The reason is that a custom Sequence class typically has many dependencies (which makes test setup too long and complex), and at the same time it provides very little functionality on top of the injected sequence actions. Bugs are much more likely to caused by the way how the real sequence action implementations interact together (which is not covered by unit tests), instead of the sequence code itself (which is the only thing covered).
```
Comment (kjdelisle) [inline pages/en/lb4/Testing-Your-application.md]: Break up into multiple lines.

Comment (kjdelisle) [reply pages/en/lb4/Testing-Your-application.md]: `Bugs are much more likely to caused by the way how the real sequence action implementations interact together (which is not covered by unit tests), instead of the sequence code itself (which is the only thing covered).`

-->
`Bugs are much more likely to exist between interacting actions within a sequence (which is not covered by unit tests), instead of the sequence code itself.`

File: pages/en/lb4/Testing-Your-application.md [hunk 17]
```diff
@@ -8,39 +8,236 @@ permalink: /doc/en/lb4/Testing-your-application.html
...
+
+See [Test Sequence customizations](#test-sequence-customizations) in [Acceptance Testing](#acceptance-testing).
+
+### Unit-test your Services
```
Comment (kjdelisle) [inline pages/en/lb4/Testing-Your-application.md]: `Unit-test your Services` -> `Unit Testing Services`

File: pages/en/lb4/Testing-Your-application.md [hunk 18]
```diff
@@ -8,39 +8,236 @@ permalink: /doc/en/lb4/Testing-your-application.html
...
+
+To be done. The initial Beta release does not include Services as a first-class feature. The following GitHub issues are tracking the related work:
+
+ - Define services to represent interactions with REST APIs, SOAP Web Services, gRPC services, and more: [#522](https://github.com/strongloop/loopback-next/issues/522)
+ - Guide: Services [#451](https://github.com/strongloop/loopback-next/issues/451)
+
+## Integration testing
```
Comment (kjdelisle) [inline pages/en/lb4/Testing-Your-application.md]: `Integration Testing`

File: pages/en/lb4/Testing-Your-application.md [hunk 19]
```diff
@@ -8,39 +8,236 @@ permalink: /doc/en/lb4/Testing-your-application.html
...
+
+Integration tests are considered as "white-box" too. They use "inside-out" approach that tests how multiple units work together or with external services. Test doubles may be used to isolate tested units from external variables/state that's not part of the tested scenario.
```
Comment (kjdelisle) [inline pages/en/lb4/Testing-Your-application.md]: `Integration tests are considered as "white-box" too.` --> `Integration tests are also considered "white-box" tests.`

Comment (kjdelisle) [reply pages/en/lb4/Testing-Your-application.md]: `They use "inside-out" approach that tests how multiple units work together or with external services.` -> `They use the "inside-out" approach, which tests how multiple units work together or with external services.`

File: pages/en/lb4/Testing-Your-application.md [hunk 20]
```diff
@@ -8,39 +8,236 @@ permalink: /doc/en/lb4/Testing-your-application.html
...
+
+### Use test data builders
```
Comment (kjdelisle) [inline pages/en/lb4/Testing-Your-application.md]: `Using Test Data Builders`

File: pages/en/lb4/Testing-Your-application.md [hunk 21]
```diff
@@ -8,39 +8,236 @@ permalink: /doc/en/lb4/Testing-your-application.html
...
+
+### Test your Repositories against a real database
```
Comment (kjdelisle) [inline pages/en/lb4/Testing-Your-application.md]: `Testing with Repositories and Real Data`

File: pages/en/lb4/Testing-Your-application.md [hunk 22]
```diff
@@ -8,39 +8,236 @@ permalink: /doc/en/lb4/Testing-your-application.html
...
+
+TODO: show how to test a custom repository method.
+
+### Test your Controllers and Repositories together
```
Comment (kjdelisle) [inline pages/en/lb4/Testing-Your-application.md]: `Testing Controllers and Repositories` or `Integration Testing with Controllers and Repositories`

File: pages/en/lb4/Testing-Your-application.md [hunk 23]
```diff
@@ -8,39 +8,236 @@ permalink: /doc/en/lb4/Testing-your-application.html
...
+
+### Test your Services against real backends
```
Comment (kjdelisle) [inline pages/en/lb4/Testing-Your-application.md]: `Integration Testing with Services`

File: pages/en/lb4/Testing-Your-application.md [hunk 24]
```diff
@@ -8,39 +8,236 @@ permalink: /doc/en/lb4/Testing-your-application.html
...
+
+To be done, the initial Beta release does not include Services as a first-class feature.
+
 ## Acceptance testing
```
Comment (kjdelisle) [inline pages/en/lb4/Testing-Your-application.md]: `Acceptance Testing` (might as well change it while we're here)

File: pages/en/lb4/Testing-Your-application.md [hunk 25]
```diff
@@ -8,39 +8,236 @@ permalink: /doc/en/lb4/Testing-your-application.html
...
 
+### Validate your OpenAPI specification
```
Comment (kjdelisle) [inline pages/en/lb4/Testing-Your-application.md]: `Validating OpenAPI Specification Schemas`
Not too sure what to refer to them as other than schemas. If there's a better word for it, use that instead.

File: pages/en/lb4/Testing-Your-application.md [hunk 26]
```diff
@@ -8,39 +8,236 @@ permalink: /doc/en/lb4/Testing-your-application.html
...
+
+### Perform an auto-generated smoke test of your REST API
```
Comment (kjdelisle) [inline pages/en/lb4/Testing-Your-application.md]: `Auto-Generating Smoke Tests for REST APIs`

File: pages/en/lb4/Testing-Your-application.md [hunk 27]
```diff
@@ -8,39 +8,236 @@ permalink: /doc/en/lb4/Testing-your-application.html
...
+
+### Test your individual REST API endpoints
```
Comment (kjdelisle) [inline pages/en/lb4/Testing-Your-application.md]: `Testing Individual REST API Endpoints`

File: pages/en/lb4/Testing-Your-application.md [hunk 28]
```diff
@@ -8,39 +8,236 @@ permalink: /doc/en/lb4/Testing-your-application.html
...
+
+### Test Sequence customizations
```
Comment (kjdelisle) [inline pages/en/lb4/Testing-Your-application.md]: `Testing Sequence Customization`

File: pages/en/lb4/Testing-Your-application.md [hunk 29]
```diff
@@ -8,39 +8,236 @@ permalink: /doc/en/lb4/Testing-your-application.html
...
+
+
+## EVERYTHING BELOW IS LEGACY AND SHOULD BE REMOVED BEFORE LANDING
```
Comment (kjdelisle) [inline pages/en/lb4/Testing-Your-application.md]: Convert this to an HTML comment instead?

File: pages/en/lb4/Thinking-in-LoopBack.md [hunk 30]
```diff
@@ -355,6 +355,8 @@ describe('API specification', () => {
 });
 ```
 
+See [Validate your OpenAPI specification](./Testing-your-application.html#validate-your-openapi-specification) from Testing your application for more details.
```
Comment (kjdelisle) [inline pages/en/lb4/Thinking-in-LoopBack.md]: Update link based on feedback.

Comment (kjdelisle) [reply pages/en/lb4/Thinking-in-LoopBack.md]: `Testing your application` -> `testing your application`

File: pages/en/lb4/Thinking-in-LoopBack.md [hunk 31]
```diff
@@ -525,6 +527,8 @@ It's a powerful proposition to use the API spec not only for API declaration but
 
 At this point, we are ready to make these tests pass by coding up your business logic.
 
+Please refer to [Perform an auto-generated smoke test of your REST API](./Testing-your-application.html#perform-an-auto-generated-smoke-test-of-your-rest-api) from Testing your application for more details.
```
Comment (kjdelisle) [inline pages/en/lb4/Thinking-in-LoopBack.md]: `Testing` -> `testing`

File: pages/en/lb4/Thinking-in-LoopBack.md [hunk 32]
```diff
@@ -982,6 +990,8 @@ The new unit test is passing now, but our integration and acceptance tests are b
  2. Fix the acceptance test by annotating `ProductController`'s `repository` argument with `@inject('repositories.Product')`
     and binding the `ProductRepository` in the main application file where we are also binding controllers.
 
+Learn more about Controller unit testing in [Unit-test your Controllers](./Testing-your-application.html#unit-test-your-controllers) from Testing your application.
```
Comment (kjdelisle) [inline pages/en/lb4/Thinking-in-LoopBack.md]: Update link based on feedback.

File: pages/en/lb4/Thinking-in-LoopBack.md [hunk 33]
```diff
@@ -682,6 +688,8 @@ When you scroll up in the test output, you will see more information about the 4
 Unhandled error in GET /product/ink-pen: 404 Error: Controller method not found: ProductController.getDetails
 ```
 
+Learn more about acceptance testing in [Test your individual REST API endpoints](./Testing-your-application.html#test-your-individual-rest-api-endpoints) from Testing your application.
```
Comment (kjdelisle) [inline pages/en/lb4/Thinking-in-LoopBack.md]: Update link based on feedback.

**Review by kjdelisle — DISMISSED** [review]
Overall comments:
~- Capitalize almost all words in a title except for conjunctions and prepositions ("for", "of", "with", etc.)~
~- Prefer use of future tense without possession for titles:`{X} your {Y}` -> `{X}ing {Y}` (ex. `Test your foos with magic` --> `Testing Foos with Magic`)~
Ignore these, I didn't realize our style guide requires this.
- Remove random capitalization of certain words ("Test" was the word most commonly capitalized outside of the beginning of a sentence or title).
- Don't join "unit test" with a hyphen (`unit-test`).
- Avoid starting sentences with "and".

File: pages/en/lb4/Testing-Your-application.md (hunk 3 above)
Comment (kjdelisle) [reply pages/en/lb4/Testing-Your-application.md:202]: "Creating Stub Repositories"

File: pages/en/lb4/Testing-Your-application.md (hunk 1 above)
Comment (crandmck) [reply pages/en/lb4/Testing-Your-application.md:19]: Re
> Capitalize almost all words in a title except for conjunctions and prepositions...

We use sentence case for headings; cf. http://loopback.io/doc/en/contrib/Doc-style-guide.html#headings

**Review by crandmck — APPROVED** [review]
Most of Kevin's comments re wording, grammar, and style are fine.  However, I'll copy edit the article once the technical content is in place and make any remaining changes of this nature.    So if you don't have time or the inclination to do these now, I can handle them post-merge.

File: pages/en/lb4/Testing-Your-application.md [hunk 34]
```diff
@@ -8,39 +8,236 @@ permalink: /doc/en/lb4/Testing-your-application.html
 summary:
 ---
 
-## Setup
+## Overview
 
-### With LoopBack CLI
+An automated test suite is important to ensure your application works as expected, prevent regressions as new features are added and give you confidence to refactor your codebase to keep it easy to further modify.
```
Comment (b-admike) [inline pages/en/lb4/Testing-Your-application.md:13]: I suggest `are added, and give you confidence to continuously refactor your codebase.` here

File: pages/en/lb4/Testing-Your-application.md [hunk 35]
```diff
@@ -8,39 +8,236 @@ permalink: /doc/en/lb4/Testing-your-application.html
...
 {% include note.html content="Currently the LoopBack CLI does not yet support LoopBack 4.
 " %}
 
 If you are just getting started with LoopBack, use the LoopBack command-line tool (CLI)  `loopback-cli`. It's ready to use and installs simply by `npm install -g loopback-cli`. You don't need to do any extra steps for setup, and can head straight to the [next section](Testing-Your-Application#acceptance-testing.html).
 
-### Without LoopBack CLI
+### Setup testing infrastructure manually
 
-If you have an existing application you'll need to install a few packages to make everything work well together. Minimally, you will need `@loopback/testlab` and `mocha`.
+If you have an existing application you'll need to install `mocha` and `@loopback/testlab` to make everything work well together.
```
Comment (b-admike) [inline pages/en/lb4/Testing-Your-application.md:47]: We can make it more clear by adding `... to make everything work well together. Inside your application root directory, run:`

File: pages/en/lb4/Testing-Your-application.md (hunk 9 above)
Comment (b-admike) [inline pages/en/lb4/Testing-Your-application.md]: `how the stubbed method was executed`

**Comment by bajtos** [comment]
@kjdelisle @b-admike Thank you for the feedback. I'd rather leave the copy editing and style/grammar fixes to @crandmck, as he suggested. It will be more effective when he makes the edits directly, instead of having to go through your comments and apply them in all places.

Is there anything else beside my style/grammar that you would like to change or improve?

I pushed another commit with some new content, you may want to take another look.

**Comment by bajtos** [comment]
Done. @strongloop/loopback-devs PTAL at the content, please focus your review on the technical part (are my advices good/correct, the code examples easy to understand, etc.) I'll leave the formal side (style, grammar, etc.) to @crandmck to improve in a follow-up pull request. If you feel some formal problem must be fixed before this can be landed, then please make the edits yourself directly in my branch - for example using GitHub's online editor:

 - [Testing Your Application](https://github.com/strongloop/loopback.io/edit/lb4/testing-your-app/pages/en/lb4/Testing-Your-application.md?pr=/strongloop/loopback.io/pull/479)
 - [Thinking in LoopBack](https://github.com/strongloop/loopback.io/edit/lb4/testing-your-app/pages/en/lb4/Thinking-in-LoopBack.md?pr=/strongloop/loopback.io/pull/479)

> Test doubles -> test doubles

I was using capital letters to make special terms stand out, e.g. "Test doubles", "Controllers", "Repositories", etc. Feel free to change it to whatever makes more sense.

Similarly, plain-text "Testing your application" was meant as a reference to a docs page. I changed the plain-test occurrences to a hyperlink to make my intent more clear. Having said that, I am not sure if it's a good idea to have two links everywhere - one pointing to a section of "Testing your application", another pointing to the top of that docs page. Feel free to change that.

File: pages/en/lb4/Testing-Your-application.md [hunk 36]
```diff
@@ -8,676 +8,556 @@ permalink: /doc/en/lb4/Testing-your-application.html
...
-1. Write a failing acceptance test.
-1. Write just enough code to make it pass.
-1. Write a failing unit test(s).
-1. Write just enough code to make it pass.
-1. Continue writing unit tests until satisfied.
-1. Write integration tests to ensure compatibility between multiple units.
-1. Write just enough code to make it pass.
-1. Continue writing integration tests until satisfied.
-1. Run the acceptance test to ensure everything continues to work together as a whole.
+Custom sequence behavior is best tested by observing changes in behavior of affected endpoints. For example, if your sequence involves authentication step that rejects anonymous requests for certain endpoints, then you can write a test making an anonymous request to such endpoint to verify that it's correctly rejected. These tests are esssentially the same as the test verifying implementation of individual endpoints as described in the previous section.
```
Comment (bajtos) [inline pages/en/lb4/Testing-Your-application.md:561]: I am not sure if this is the best advice, especially considering our own advice in [Thinking in LoopBack](https://github.com/strongloop/loopback.io/blob/0be8c43d389c62a0122eeb29cd04f338b2387aa1/pages/en/lb4/Thinking-in-LoopBack.md#how-to-build-a-great-test-suite):

> How to create a great test suite? By thinking smaller and favoring fast, focused unit-tests over slow application-wide end-to-end tests.

On one hand, writing several end-to-end tests to verify how auth is applied to each endpoint brings quite a lot of confidence for the security of our app. On the other hand, these tests are repetitive to write and take a long time to execute.

Perhaps we should look for better ways, for example:

 - A small number of integration tests to make sure all sequence customizations are applied for HTTP requests, these tests can invoke a dummy test controller method instead of real API endpoints - we are interested in the sequence, not any specific controller method.

 - Unit test for each controller method to ensure the method is correctly decorated with the metadata used by additional sequence actions used in our custom sequence.

For example, let's say we are using `@loopback/authentication` and its `@authenticate` decorator and `authenticate()` sequence action. To test such app, we would write:

 - four integration tests: two tests for a controller method decorated with `@authenticate` (an anonymous request, an authenticated request), two tests for a controller method not decorated with `@authenticate` (again, an anonymous request, an authenticated request)

 - for each controller method, one unit test verifying that the method is (or is not) decorated with `@authenticate`

@kjdelisle thoughts?

If we agree to rewrite this section, then I am proposing to do so in a follow-up pull request, so that this big patch can be landed ASAP.

File: pages/en/lb4/Testing-Your-application.md [hunk 37]
```diff
@@ -8,676 +8,556 @@ permalink: /doc/en/lb4/Testing-your-application.html
...
-  // Shut down the application and its server instance.
-  async function stopApp() {
-    await app.stop();
-  }
+
+```ts
+// test/helpers/database.helpers.ts
+
+export async function givenEmptyDatabase() {
+  await new ProductRepository().deleteAll();
```
Comment (jannyHou) [inline pages/en/lb4/Testing-Your-application.md:86]: Would it be better to recommend using docker to setup a clean database?

- from my experience api like `deleteAll()` some times also interferes testing result/pollute db if it has bug and could not clean up the old data entirely. 

- you can also leave the container there instead of deleting it after test :)

I feel we benefit a lot after lb3 connectors all use container for testing.

Comment (bajtos) [reply pages/en/lb4/Testing-Your-application.md:86]: We are running `givenEmptyDatabase` before each test. How long does it take to restart a docker container with a database? I am concerned that cleaning the database via docker restart would make the test suite extremely slow.

> from my experience api like deleteAll() some times also interferes testing result/pollute db if it has bug and could not clean up the old data entirely.

I see your point. This can happen for example when there are foreign keys setup between tables and let' say a `Category` record cannot be deleted until all `Product` records belonging to that category are deleted too. IMO, this is a missing feature of LoopBack - we should support cascading deletes. Fortunately, this particular example has a workaround - just re-order `deleteAll` statements so that products are deleted before categories.

I think we shouldn't be recommending our users suboptimal solutions just to avoid shortcomings in LoopBack and/or the way how the users designed their app & data schema. What I am proposing instead: let's use this as an opportunity to identify missing gaps and either explain users how to work around them, or work together to fix these issues.

@jannyHou What do you think, should I add a short paragraph explaining how to clear database when there are relationships between tables enforced by the DB engine, like I shown in my first paragraph?

File: pages/en/lb4/Testing-Your-application.md (hunk 36 above)
Comment (bajtos) [reply pages/en/lb4/Testing-Your-application.md:561]: Opened a p2 issue to keep track of my suggestion: https://github.com/strongloop/loopback-next/issues/648

File: pages/en/lb4/Testing-Your-application.md (hunk 37 above)
Comment (jannyHou) [reply pages/en/lb4/Testing-Your-application.md:86]: @bajtos yeah the time is a problem for starting a container...

I think we can add more after figure out a plan for persistence story, now your content is good enough, going into too much details might be off topic.
